File: frontend/lib/api/ai.ts
import { ApiClient } from "./client";

export interface MaterialAnalysisResult {
  material: string;
  category: string;
  confidence: number;
  quality: string;
  recyclable: boolean;
  estimated_price_per_kg: number;
  tips?: string;
}

export interface GeneratedListingResult {
  title: string;
  description: string;
  category: string;
  material: string;
  quality: string;
  estimated_price_min: number;
  estimated_price_max: number;
  estimated_price_display: string;
}

export interface EcoBotResult {
  answer: string;
  suggestions: string[];
}

export const aiApi = {
  analyzeMaterial: async (imageUrl: string): Promise<MaterialAnalysisResult> => {
    const res = await ApiClient.post<MaterialAnalysisResult>("/api/v1/ai/analyze-material", {
      image_url: imageUrl,
    });
    if (!res.success || !res.data) throw new Error(res.message || "Failed to analyze material");
    return res.data;
  },

  generateListing: async (data: {
    material: string;
    category: string;
    quality: string;
    quantity: number;
    unit?: string;
  }): Promise<GeneratedListingResult> => {
    const res = await ApiClient.post<GeneratedListingResult>("/api/v1/ai/generate-listing", data);
    if (!res.success || !res.data) throw new Error(res.message || "Failed to generate listing");
    return res.data;
  },

  queryEcoBot: async (
    query: string,
    history: Array<{ role: string; content: string }> = []
  ): Promise<EcoBotResult> => {
    const res = await ApiClient.post<EcoBotResult>("/api/v1/ai/ecobot", { query, history });
    if (!res.success || !res.data) throw new Error(res.message || "Failed to query EcoBot");
    return res.data;
  },
};

File: backend/app/api/v1/ai.py
from typing import Optional
from fastapi import APIRouter, Depends
from backend.app.core.security import get_optional_current_user
from backend.app.models.user import User
from backend.app.schemas.common import APIResponse
from backend.app.schemas.ai import (
    MaterialAnalyzeRequest,
    MaterialAnalyzeResponse,
    ListingGenerateRequest,
    ListingGenerateResponse,
    EcoBotRequest,
    EcoBotResponse,
)
from backend.app.services.ai_service import ai_service

router = APIRouter(prefix="/ai", tags=["AI Services"])


@router.post("/analyze-material", response_model=APIResponse[MaterialAnalyzeResponse])
async def analyze_material(
    payload: MaterialAnalyzeRequest,
    current_user: Optional[User] = Depends(get_optional_current_user),
):
    """Analyze an uploaded scrap image and detect recyclable material type."""
    result = await ai_service.identify_material(image_url=payload.image_url)
    return APIResponse.respond(
        data=MaterialAnalyzeResponse(**result),
        message="AI material identification complete",
    )


@router.post("/generate-listing", response_model=APIResponse[ListingGenerateResponse])
async def generate_listing_content(
    payload: ListingGenerateRequest,
    current_user: Optional[User] = Depends(get_optional_current_user),
):
    """Generate listing title, description, and price estimation from detected material."""
    result = await ai_service.generate_listing(
        material=payload.material,
        category=payload.category,
        quality=payload.quality,
        quantity=payload.quantity,
        unit=payload.unit,
    )
    return APIResponse.respond(
        data=ListingGenerateResponse(**result),
        message="Listing generated",
    )


@router.post("/ecobot", response_model=APIResponse[EcoBotResponse])
async def query_ecobot(
    payload: EcoBotRequest,
    current_user: Optional[User] = Depends(get_optional_current_user),
):
    """Ask EcoBot questions about recycling, disposal methods, and pickup procedures."""
    result = await ai_service.query_ecobot(query=payload.query, history=payload.history)
    return APIResponse.respond(
        data=EcoBotResponse(**result),
        message="EcoBot response generated",
    )
